pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyTest.java

```diff
@@ -429,6 +436,63 @@ public void testGetClientVersion() throws Exception {
                 .get(0).getClientVersion(), String.format("Pulsar-Java-v%s", PulsarVersion.getVersion()));
     }
 
+    @DataProvider
+    public Object[][] booleanValues() {
+        return new Object[][]{
+                {true},
+                {false}
+        };
+    }
+
+    @Test(dataProvider = "booleanValues")
+    public void testConnectedWithClientSideFeatures(boolean supported) throws Exception {
+        final String topic = BrokerTestUtil.newUniqueName("persistent://public/default/tp");
+        admin.topics().createNonPartitionedTopic(topic);
+
+        // Create a client as a old version, which does not support "supportsReplDedupByLidAndEid".
+        ClientBuilderImpl clientBuilder2 =
+                (ClientBuilderImpl) PulsarClient.builder().serviceUrl(proxyService.getServiceUrl());
+        PulsarClientImpl injectedClient = InjectedClientCnxClientBuilder.create(clientBuilder2,
+            (conf, eventLoopGroup) -> {
+                return new ClientCnx(InstrumentProvider.NOOP, conf, eventLoopGroup) {
+
+                    @Override
+                    protected ByteBuf newConnectCommand() throws Exception {
+                        authenticationDataProvider = authentication.getAuthData(remoteHostName);
+                        AuthData authData = authenticationDataProvider.authenticate(AuthData.INIT_AUTH_DATA);
+                        BaseCommand cmd =
+                                Commands.newConnectWithoutSerialize(authentication.getAuthMethodName(), authData,
+                                        this.protocolVersion, clientVersion, proxyToTargetBrokerAddress,
+                                        null, null, null, null, null);
+                        cmd.getConnect().getFeatureFlags().setSupportsAuthRefresh(supported);
+                        cmd.getConnect().getFeatureFlags().setSupportsBrokerEntryMetadata(supported);
+                        cmd.getConnect().getFeatureFlags().setSupportsPartialProducer(supported);
+                        cmd.getConnect().getFeatureFlags().setSupportsTopicWatchers(supported);
+                        cmd.getConnect().getFeatureFlags().setSupportsReplDedupByLidAndEid(supported);
+                        cmd.getConnect().getFeatureFlags()
```

Comment:
using a local variable for `cmd.getConnect()` would make this test code more readable.

Reply:
done